modules/phase_field/src/auxkernels/EulerAngleProvider2RGBAux.C

```diff
@@ -50,32 +52,70 @@ EulerAngleProvider2RGBAux::EulerAngleProvider2RGBAux(const InputParameters & par
     _grain_tracker(getUserObject<GrainTracker>("grain_tracker")),
     _no_grain_color(getParam<Point>("no_grain_color"))
 {
+  if (isParamValid("phase"))
+    _phase = getParam<unsigned int>("phase");
+}
+
+unsigned int
+EulerAngleProvider2RGBAux::getNumGrains() const
+{
+  if (isParamValid("phase"))
+  {
+    const EBSDReader * euler = dynamic_cast<const EBSDReader *>(&_euler);
```

Comment:
Parameter checking fetching and casting should be done in the constructor.

Reply:
Are you saying that the `dynamic_cast` should also be moved to the constructor?

Reply:
@dschwen & @permcody,  
I added the dynamic cast to the constructor as follows: `_ebsd_reader(dynamic_cast<const EBSDReader &> (_euler)),`. It compiles but a few tests (EulerAngleProvider2RGBAux_bicrystal.i, EulerAngle2RGBAction.i, and random_grain_orientation.i) are failing with `*** ERROR ***  std::bad_cast`.

In order to apply the dynamic cast to only the phase case, I tried this:
`_ebsd_reader(isParamValid("phase") ? dynamic_cast<const EBSDReader &> (_euler) : NULL),`. Of course this failed because you can't have a null reference.  Is there another way to do this?

Reply:
You'll have to keep (leave) _ebsd_reader as a pointer:
```
dynamic_cast<const EBSDRead *>(&_euler)
```

Reply:
That results in:  `error: reference to type 'const EBSDReader' could not bind to an rvalue of type 'const EBSDReader *'`

Reply:
Not sure why you are seeing that - do we have a name clash in there? That almost identical piece of code is already checked in.